src/components/hack_forms/questions/RegQuestionOptionEditor.js

```diff
@@ -0,0 +1,155 @@
+   * pre-existing option, we simply replace the old value and pass
+   * it to the parent. However, if we edited a new option, we need
+   * to add the new option.
+   */
+  const handleEdit = (event, index) => {
+    const newOptions = [...props.options];
+    if (index === props.options.length) {
+      newOptions.push(event.target.value);
+    } else {
+      newOptions[index] = event.target.value;
+    }
+    props.setRegOptions(newOptions);
+  };
+
+  /**
+   * Handles when an option's icon is clicked. If it is the last
+   * option (i.e., the "add new" option), it adds a new option.
+   * Otherwise, it deletes the option (in accordance with the icon's
+   * appearance: delete vs add).
+   */
+  const handleIconClicked = index => {
+    const newOptions = [...props.options];
+    if (index === props.options.length) {
+      // Add new option
+      newOptions.push("");
+    } else {
+      // Delete the option
+      newOptions.splice(index, 1);
+    }
+    props.setRegOptions(newOptions);
+  };
+
+  /** Creates a new text field for the option. */
+  const createTextField = index => {
+    return (
+      <TextField
+        className={classes.textField}
+        id={`option-${index + 1}`}
+        name={`option-${index + 1}`}
+        label={`Option ${index + 1}`}
+        value={index < props.options.length ? props.options[index] : ""}
+        onChange={event => handleEdit(event, index)}
+      />
+    );
+  };
+
+  /** Gets the icon to display for the option (add or delete, depending). */
+  const getIcon = index => {
+    if (index === props.options.length) {
+      return <AddCircleIcon fontSize="small" />;
+    } else {
+      return <RemoveCircleIcon fontSize="small" />;
+    }
+  };
+
+  /**
+   * Creates a list of checkbox options with a series of checkboxes. There
+   * is one more option visible than the actual number that exist; the last
+   * one is there for creating new ones.
+   */
+  const createCKOptions = () => {
+    return (
+      <FormGroup>
+        {[...props.options, ""].map((option, idx) => (
+          <div className={classes.formItem} key={idx}>
+            <Checkbox checked={false} />
+            {createTextField(idx)}
+            <IconButton onClick={() => handleIconClicked(idx)}>
+              {getIcon(idx)}
+            </IconButton>
+          </div>
+        ))}
+      </FormGroup>
+    );
+  };
+
+  /**
+   * Creates a list of radio button options with a series of radio buttons.
+   * There is one extra option visible, which is for adding more options.
+   */
+  const createRDOptions = () => {
+    return (
+      <RadioGroup>
+        {[...props.options, ""].map((option, idx) => (
+          <div className={classes.formItem} key={idx}>
+            <Radio value={false} />
+            {createTextField(idx)}
+            <IconButton onClick={() => handleIconClicked(idx)}>
+              {getIcon(idx)}
+            </IconButton>
+          </div>
+        ))}
```

Comment:
What's happening here with the empty string? I'm not sure I understand.

Reply:
Ha! Yeah, it's a bit strange. The empty string simply adds an extra element to the array. So, we create a text field for every option PLUS one extra text field which is used for adding an extra option. Once that option is edited, the option is added to the array `props.options` and another blank option is added to the bottom for adding.

Example: if `props.options = ["bananas", "grapes", "oranges"]`, then we show four options on the screen: `["bananas", "grapes", "oranges", ""]` where the last option is for adding another option (this component is for the admin to choose what options are available for a question).